### PR TITLE
ci: automated release workflow

### PR DESCRIPTION
**Pull request by jthegedus**
# Summary

Automated release **PRs** created by [`release-please`](https://github.com/google-github-actions/release-please-action) which uses [`conventional commits`](https://www.conventionalcommits.org/en/v1.0.0/) to determine the semver release version. **Merge the PR to create the release!**

This removes all manual steps to release making it highly repeatable, actionable by any maintainer, and actionable without a workstation or git (I can merge from my phone while out and about).

A `release-please` PR will:
- look at git history since last release
- analyze commit messages
- determine the semver change
- update the changelog
- update the version number
- if new commits are made to the default branch, the PR is updated accordingly, recalculating the above
- on merge, creates:
  - the Git Tag
  - the GitHub Release

Closes: #942

## Details

- `.github/workflows/release.yml`: release-please workflow which creates release PRs
- `.github/workflows/semantic-pr.yml`: enforces PR titles (since the title is used on squash merge) are Conventional Commit compliant
- `.github/workflows/docs.yml`: updates the asdf version in the install instructions when a new release is published (replacing `release/tag.sh` functionality). *This workflow will likely be superseded by our static docs site generator once I get to that.*
- replace `VERSION` file with `version.txt` as release-please currently requires use of this file

### Release PRs

A release PR looks like this:

image

This shows the new additions to the Changelog. This image is taken from my fork - https://github.com/jthegedus/asdf/pull/1

image

Changelog PR numbers and SHAs are actual links too!

### Documentation Version Update

I ran the workflow on my fork to test:

image

- GitHub Action execution: https://github.com/jthegedus/asdf/runs/3006470600?check_suite_focus=true
- Result commit: https://github.com/jthegedus/asdf/commit/3f2fd1e9bec35493b9d673dfff96b5326e1c919b

### What is lost

From what I understand Release Candidates are not possible using [`release-please`](https://github.com/google-github-actions/release-please-action), I may be mistaken. However useful there are other solutions to RCs, our support for `asdf update --head` being one but also enabling feature flags with a toggle command like `asdf --release-candidate` which could set a flag and opt the user into other code paths.

Nonetheless, I believe the benefits of the rest of this system outweigh the loss of RCs.

### Additional info

I have been using [`release-please`](https://github.com/google-github-actions/release-please-action) on my own projects the past few months ([`asdf-gcloud`](https://github.com/jthegedus/asdf-gcloud), [`asdf-firebase`](https://github.com/jthegedus/asdf-firebase)), it has been great. Happy to answer any questions.

@Stratus3D I appreciate the work you have put into releases and the `release/tag.sh` script over the years. Especially the difficulty with crafting the changelog etc in that script. :pray:

### CODE REVIEW

**Comment by Stratus3D** [comment]
Thanks @jthegedus ! This looks promising. I haven't reviewed all the code yet but will do so soon.

**Review by Stratus3D — APPROVED** [review]
Can you explain what this does in `release.yml`?

``` release.yml
bump-minor-pre-major: true
```

Also, I like to be able to run everything locally that GitHub also runs (I don't always have access to the internet), do you know how I would link a PR title/Git commit message locally like the `amannn/action-semantic-pull-request` does on Github?

The changes here look great, assuming you can provide me with a way to validate my commit messages locally I approve 👍

**Comment by jthegedus** [comment]
`release-please` relies on [Conventional Commit messages](https://www.conventionalcommits.org/), with:

- `fix`: which represents bug fixes, and correlates to a SemVer **patch**.
- `feat`: which represents a new feature, and correlates to a SemVer **minor**.
- `<type>!`: which represent a breaking change (indicated by the !) and will result in a SemVer **major**.

All Conventional Commit types are: `feat`, `fix`, `docs`, `style`, `refactor`, `perf`, `test`, `build`, `ci`, `chore`, `revert`.

Our current project is `0.y.z`. By definition of the Semver spec:

>Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

Given this, we should still use `!` to indicate breaking changes, however we don't want to our next release to be tagged as `1.0.0`.

```yaml
bump-minor-pre-major: true
```

will prevent `release-please` from creating `1.0.0`, instead creating a new **minor** for any breaking change.

When we want to tag `1.0.0`, we remove `bump-minor-pre-major` (it's default is `false`) and `release-please` will then use breaking changes (`!`) to bump the semver **major**.

---

>Also, I like to be able to run everything locally that GitHub also runs (I don't always have access to the internet), do you know how I would link a PR title/Git commit message locally like the `amannn/action-semantic-pull-request` does on Github?

That is fair enough.

`amannn/action-semantic-pull-request` is just a failsafe. It ensures PR titles match [Conventional Commit messages](https://www.conventionalcommits.org/) standards, thus ensuring `release-please` creates the correct changelog & version bump.

*It is intended for PR titles only*. However `validateSingleCommit: true` is set because

>When using "Squash and merge" on a PR with only one commit, GitHub will suggest using that commit message instead of the PR title for the merge commit, and it's easy to commit this by mistake. Enable this option to also validate the commit message for one commit PRs.

There are many solns to this, the most simple being push a second commit so the PR title is used instead of commit msg. Not a very nice soln, but much easier than getting every contributor to install a tool linting commits for [Conventional Commit](https://www.conventionalcommits.org/) style.

I have considered Git Hooks to perform these types of validations, but that seems heavy handed. My recommendations for tools to locally run are [`husky`](https://github.com/typicode/husky) & [`commit-lint`](https://commitlint.js.org/), but both rely on the Node.js ecosystem (I have thought about making a Husky `asdf` plugin which removes the Node.js aspects for non-Node.js projects, but I digress)

Do you think a 2nd commit is an acceptable compromise?

**Comment by jthegedus** [comment]
I will update the contributing guide to include a description of this new workflow & requirements on the PR title.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: GoogleCloudPlatform/release-please-action@v2
+        name: create release
+        with:
+          release-type: simple
+          bump-minor-pre-major: true # remove this to enable breaking changes causing 1.0.0 tag
```
Comment (jthegedus) [inline .github/workflows/release.yml:16]: I have documented this setting for future contributors

**Review by Stratus3D — APPROVED** [review]
> All that is to say, there's no direct way to exec amannn/action-semantic-pull-request locally, though it runs on PR titles which only exist in GitHub. Do you think a 2nd commit is an acceptable compromise for single-commit PRs?

I'm not sure what you mean, but since I just setup `commitlint` for local development I think what you have here is fine. Thanks for answering all my questions!

**Comment by jthegedus** [comment]
Linting a commit is only required when a PR has a **single commit** as the "squash & merge" GitHub PR strategy uses the commit name as the merge message, not PR title.

**Comment by jthegedus** [comment]
Thanks for reviewing

**Pull request merged**